### ISSUE TITLE
Can we tune faster?

### ISSUE THREAD

**Issue by kodonnell** [CNugteren/CLBlast#275]
Faster tuning means:

- faster testing/development
- easier to consider a larger search space
- happy users

Some ideas:

### Timeout

For example, while tuning xgemm, the fastest kernels run in under a millisecond - while others take over ten seconds. In this case its 'obvious' that as soon as a kernel has taken e.g. 2ms, it's not going to be the fastest, so you might as well skip it. It'd be great to be able to:

- set a specific timeout (e.g. 50ms in this case), and
- have the tuner keep track of the fastest result so far, and update the timeout as it goes, as there's no point testing slower ones.

I'd like to think this isn't too hard ... though doing it nicely might be. E.g. if you can't interrupt kernels nicely, then you could always just run a separate thread (?) for each config, and kill the whole thread. (Or, worse: use bash to loop through and do this, etc.)

### Bail if the first run(s) are too slow

If I set `runs=10`, but the first run is already taking too long, then why not give up instead of running the other 9? It'd need some care (as there is some variability etc. between runs, and warmup to consider) but it could be an easy win.

**Comment by CNugteren** [comment]
A bit of a late reply, but I see a couple of problems with your first suggestion to interrupt slower kernels:

1. Typically running the kernel itself is the least amount of the total tuning time. Other parts include compilation time (typically around half a second but sometimes even multiple seconds), copying to the host, and correctness checking (last 2 points could be optimised perhaps). However this is different per platform of course, especially when going to larger matrices kernel run-time becomes important as well.  

2. I don't think it is possible to interrupt OpenCL kernels halfway based on a time-out. If you have any reference, let me know.

3. The tuner doesn't only care about the best result. Also other valid results can contribute to a architecture-level, vendor-level, or global default. Furthermore, all tuning results are stored in a database to be potentially be used to learn predictive models in the future.


I see a bit more value in your second idea: bail out if the first few runs are slow. However, I would still suggest to go at least 2 times. And 2 is actually already the default setting for the slower GEMM kernels,so also not much to gain here it seems...

**Comment by kodonnell** [comment]
> Typically running the kernel itself is the least amount of the total tuning time. 

For quick kernels, yes. Some of the kernels were reporting 30 second runtimes, if I remember correctly, and I presume in these cases they're much longer than compilation. Anyway, it's these ones that I'd really want to get rid of.

> I don't think it is possible to interrupt OpenCL kernels halfway ...

I have no idea. A quick look reveals some systems (windows?) have a timeout on any kernel - but I haven't seen any software suggestions, aside from killing the command queue (though I'm not sure that'd work). You could maybe do other things like adapting the kernels to monitor a global 'keep running' variable, or maybe try other nasty things like messing with the global inputs or some such. Anyway, just guessing.

Question - if I `ctrl + c` the tuner, does that kill the kernels on the GPU? If so, then there's at least a feasible (but ugly) solution.

> The tuner doesn't only care about the best result. Also other valid results can contribute to a architecture-level, vendor-level, or global default. Furthermore, all tuning results are stored in a database to be potentially be used to learn predictive models in the future.

Agreed, though I imagine it'll be rather hard to find a predictive model that beats an exhaustive grid search. I guess that's where I'm coming from - if the user wants to tune for the fastest kernel, then you can either have a complex algorithm for doing it (big database of other tunings, predictive models, etc.) which is not guaranteed to return the fastest result - and it might still take a lot of time. The alternative is much simpler: just exhaust the search space, but very quickly (by bailing out of anything that isn't the fastest). Of course, it depends on the search space, but if I remember correctly, for GEMM it was around a hundred thousand or so, so not particularly big.

**Comment by CNugteren** [comment]
Yes, there might be some cases where it could be beneficial, although I don't think this will speed-up tuning by a significant amount for most use-cases. But anyway, I don't see a feasible solution at all (see point 2). A quick search indeed [confirms this](https://stackoverflow.com/questions/9039155/is-there-any-way-to-stop-opencl-kernel-from-execution#9039650). So unless someone has a solution, I'm afraid this is not the way to make tuning faster.

**Comment by kodonnell** [comment]
> A quick search ...

But that does hint at out a solution (though, as above, it'd involve rewriting the kernels).

> So unless someone has a solution, I'm afraid this is not the way to make tuning faster.

Agreed.

**Comment by kodonnell** [comment]
Also - I did originally intend this issue to include other ideas, not just my own, so I'll leave this open. (You're welcome to close it though.)

**Comment by CNugteren** [comment]
> But that does hint at out a solution (though, as above, it'd involve rewriting the kernels).

That seems very hacky and very involved, will not work on all devices (requires multiple concurrent kernels) and actually might mess up the time measurement of the kernel itself!

> I did originally intend this issue to include other ideas

OK, makes sense. Yes, any idea is welcome. I think the most viable thing right now to work on is to support zero-copy buffers to save copying overhead on devices such as Intel GPUs or embedded devices.

**Comment by CNugteren** [comment]
You make some good suggestions. However, keep in mind that in the majority of cases the kernel compilation time is the bottleneck, so spending effort to optimize other parts is a waste of resources. Indeed, compilation could be parallelized, but that would be some non-trivial amount of work and would complicate the code. I'm open to accepting PRs on parallel compilation if someone wants to make an attempt.

**Comment by EJain-Dev** [comment]
Would it possible through some effort to compile multiple of the same kernel in 1 program to reduce overhead?

**Comment by EJain-Dev** [comment]
Pull request is complete (#580) and compiles fine on my system but fails with appveyor on initializer list at line 187. Although, appveyor is does not say required, is this something to fix?
